### PR TITLE
[Search] name in search result clickable

### PR DESCRIPTION
**Pull request by emmaindal**
This PR makes it possible to navigate to the service catalog page of an entity by clicking on the name. It also changes the column name from "component id" to "name" to be more consistent in terminology across Backstage. 

closes: #3345

Before: 
Screen Shot 2020-11-19 at 12 49 54 PM

After:
Screen Shot 2020-11-19 at 12 48 29 PM

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)

### CODE REVIEW

File: plugins/search/src/components/SearchResult/SearchResult.tsx [hunk 1]
```diff
@@ -65,9 +66,14 @@ type Filters = {
 // TODO: move out column to make the search result component more generic
 const columns: TableColumn[] = [
   {
-    title: 'Component Id',
+    title: 'Name',
     field: 'name',
     highlight: true,
+    render: (entity: Partial<Result>) => (
+      <Link to={`/catalog/${entity.namespace}/${entity.kind}/${entity.name}`}>
```
Comment (iamEAP) [inline plugins/search/src/components/SearchResult/SearchResult.tsx:73]: What do you think of encapsulating the URL/path in the `apis.ts` file above, rather than here?  So instead of adding `namespace` and pulling it out here, we add `url` and set `<Link to={entity.url} ... />`

Comment (emmaindal) [inline plugins/search/src/components/SearchResult/SearchResult.tsx:73]: oh actually, just realized [namespace can be optional](https://backstage.io/docs/architecture-decisions/adrs-adr002#namespace) now, so need to change that here

Comment (emmaindal) [reply plugins/search/src/components/SearchResult/SearchResult.tsx:73]: like it!

Comment (hooloovooo) [reply plugins/search/src/components/SearchResult/SearchResult.tsx:73]: I don't think they're supposed to be optional in the URL, and should be `default` if the entity doesn't have a namespace. Not sure that's how it's implemented in the catalog though. https://backstage.io/docs/architecture-decisions/adrs-adr009#entity-references-in-urls

Comment (iamEAP) [reply plugins/search/src/components/SearchResult/SearchResult.tsx:73]: Would be cool if entities had methods like `getBaseUrl()` or something so we didn't have to fiddle with / know these kinds of things.

Comment (emmaindal) [reply plugins/search/src/components/SearchResult/SearchResult.tsx]: ^^ really good idea

Comment (emmaindal) [reply plugins/search/src/components/SearchResult/SearchResult.tsx]: @iamEAP done

Comment (OrkoHunter) [reply plugins/search/src/components/SearchResult/SearchResult.tsx]: I recently learnt about `getBaseUrl` in backend plugins using [`discovery`](https://github.com/backstage/backstage/tree/master/packages/backend-common/src/discovery) APIs. I am guessing there is [`DiscoveryApi`](https://github.com/backstage/backstage/blob/master/packages/core-api/src/apis/implementations/DiscoveryApi/UrlPatternDiscovery.ts) for frontend too.

Comment (emmaindal) [reply plugins/search/src/components/SearchResult/SearchResult.tsx]: Hmmm interesting, so I think here we want to get a url for each entity and not actually a plugin, the DiscoveryApi seems to be responsible for constructing URLs for each plugin. 🤔 Or I could be wrong here

Comment (freben) [reply plugins/search/src/components/SearchResult/SearchResult.tsx]: Definitely not ON entities - but yes. We are right now settling how routing and page composition is to work in the final non-alpha product, and this sounds like one of those places where route refs come into play.

Scarce old info here: https://backstage.io/docs/reference/createPlugin-router

The route refs will probably remain (still working on the details) and you'd essentially be using a facility to resolve the target URL using them. There would be something like this one https://github.com/backstage/backstage/blob/master/plugins/catalog/src/plugin.ts#L25 but for the entity landing page.

Comment (freben) [reply plugins/search/src/components/SearchResult/SearchResult.tsx]: To get the three parts out of an entity including the default, you can use this https://github.com/backstage/backstage/blob/master/packages/catalog-model/src/entity/ref.ts#L28

Comment (emmaindal) [reply plugins/search/src/components/SearchResult/SearchResult.tsx]: Hmmm @freben I don't know if it makes sense here to add catalog model as a dependency only because I want the default fallback 🤔 I already have access to name, namespace and kind, [see changes.](https://github.com/backstage/backstage/pull/3349/files#diff-7dd9aeca72a846e6cd397452886389314a602a138c39d8d545c1bb49c45e25e9R45)

Comment (freben) [reply plugins/search/src/components/SearchResult/SearchResult.tsx]: If it makes it easier to stomach, that package is very special - it is ultra light weight on its own depencencies and is special made to be imported everywhere by both frontend and backend packages because it holds the "master" types to deal with entities for everybody. So if you use the Entity type, for example, you are actually already using it from that one.

Comment (freben) [reply plugins/search/src/components/SearchResult/SearchResult.tsx]: Also that string `'default'` exists as an exported const from the same package for the same reasons :)

Comment (emmaindal) [reply plugins/search/src/components/SearchResult/SearchResult.tsx]: Hey @freben thanks for response, I don't use the Entity type yet but I probably should, so that gave me at least one more reason to use it. 🎉

**Comment by emmaindal** [comment]
@iamEAP @freben have updated things related to your comments so when you have time, take a look again please!

File: plugins/search/src/apis.ts [hunk 2]
```diff
@@ -35,12 +37,19 @@ class SearchApi {
...
-      name: result.metadata.name,
-      description: result.metadata.description,
-      owner: result.spec.owner,
-      kind: result.kind,
-      lifecycle: result.spec.lifecycle,
+    return entities.items.map((entity: Entity) => ({
+      name: entity.metadata.name,
+      description: entity.metadata.description || 'No description',
+      owner:
+        typeof entity.spec?.owner === 'string' ? entity.spec?.owner : 'unknown',
```
Comment (freben) [inline plugins/search/src/apis.ts:44]: In general when making a fallback like this, an idiomatic way is to use `??` when you know that it's always an undefined that you are replacing, or the good old `||` if you want to be liberal in replacing everything falsy.

Here you know that the type is always going to be a string, so what you actually wanted was probably the former, but could also use the latter to cover empty strings (which should be forbidden, but ...).

TL:DR....: `owner: entity.spec?.owner ?? 'unknown'` probably fits well here, to avoid repetition.

File: plugins/search/src/apis.ts [hunk 3]
```diff
@@ -35,12 +37,19 @@ class SearchApi {
...
+      kind: entity.kind,
+      lifecycle:
+        typeof entity.spec?.lifecycle === 'string'
+          ? entity.spec?.lifecycle
+          : 'unknown',
```
Comment (freben) [inline plugins/search/src/apis.ts:49]: same here as above.

One more side note for a later refactor - it seems to me that you will end up rendering mixed result types together in different rows. You actually don't want the string "unknown" here right? For entities that don't have a lifecycle (validly! because it just isn't a thing for them!) you would just want to show an empty cell instead.

TL:DR....: Maybe this should actually just be `lifecycle: entity.spec?.lifecycle` and let it be undefined if that happens to be the case.

File: plugins/search/src/apis.ts [hunk 4]
```diff
@@ -35,12 +37,19 @@ class SearchApi {
...
+      url: `/catalog/${entity.metadata.namespace || ENTITY_DEFAULT_NAMESPACE}/${
+        entity.kind
+      }/${entity.metadata.name}`,
```
Comment (freben) [inline plugins/search/src/apis.ts:52]: OK now I'm being even more annoying ... :) But to appease @stefanalund , you may actually want to lowercase the namespace and kind. The URLs look a whole ton nicer as `/catalog/default/component/blah` than  `/catalog/default/Component/blah`. :) And it's intentionally case insensitive there.

File: plugins/search/src/apis.ts (hunk 2 above)
Comment (emmaindal) [reply plugins/search/src/apis.ts:44]: Before I had empty columns as you mentioned below, but the problem I experienced here was that I'm seeing owners can be "string | number | boolean | JsonObject | JsonArray" 🤔  which was the reason I had to check that it was a string here

Comment (emmaindal) [reply plugins/search/src/apis.ts:44]: It might be something I do wrong with types that can be solved instead

Comment (freben) [reply plugins/search/src/apis.ts:44]: Ohh I see, it's because you're looking at the generic Entity type and it says that "in general, a spec can contain anything that json accepts", while the actual subtypes narrow it down to "these actual fields are the ones that actually exist".

Yeah OK. So `typeof entity.spec?.owner === 'string' ? entity.spec.owner : undefined` could perhaps make sense? The alternative would be to (perhaps in a separate function) switch on the `kind`, cast to the right type for that, and read. But that may not buy us much here.

**Review by iamEAP — APPROVED** [review]
I love walking into a squeaky clean PR after the review dust has settled.

:shipit:

File: plugins/search/src/apis.ts [hunk 5]
```diff
@@ -35,12 +37,18 @@ class SearchApi {
...
-    return entities.items.map((result: any) => ({
-      name: result.metadata.name,
-      description: result.metadata.description,
-      owner: result.spec.owner,
-      kind: result.kind,
-      lifecycle: result.spec.lifecycle,
+    return entities.items.map((entity: Entity) => ({
+      name: entity.metadata.name,
+      description: entity.metadata.description || 'No description',
+      owner: typeof entity.spec?.owner ?? undefined,
```
Comment (freben) [inline plugins/search/src/apis.ts:43]: You accidentally deleted a bit much here

Comment (emmaindal) [reply plugins/search/src/apis.ts:43]: ups, and I also did not take away the fallback for description which I planned to

Comment (emmaindal) [reply plugins/search/src/apis.ts]: fixed!
